sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts

```diff
@@ -290,50 +290,61 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
    * Updates assessment status and save workflow data ensuring
    * that the data will be updated only once.
    * @param assessmentId updated assessment.
-   * @param workflowData workflow data to be saved.
    * @param entityManager used to execute the commands in the same transaction.
+   * @param options options to be updated.
+   * - `workflowData` data to be saved in the workflowData field.
    * @returns true if the update was executed or false in case the data was already present.
    */
-  async updateAssessmentStatusAndSaveWorkflowData(
+  async updateAssessmentWrapUpData(
     assessmentId: number,
-    workflowData: WorkflowData,
     entityManager: EntityManager,
+    options?: { workflowData?: WorkflowData },
   ): Promise<boolean> {
     const studentAssessmentRepo =
       entityManager.getRepository(StudentAssessment);
-    const auditUser = this.systemUsersService.systemUser;
-    const now = new Date();
     const studentAssessment = await studentAssessmentRepo.findOne({
-      select: { studentAssessmentStatus: true, workflowData: true as unknown },
+      select: { studentAssessmentStatus: true },
       where: { id: assessmentId },
     });
-    if (studentAssessment.workflowData) {
-      // If the workflow data was already updated no further updates are needed.
-      return false;
+    const workflowIsCancelled = [
+      StudentAssessmentStatus.CancellationRequested,
+      StudentAssessmentStatus.CancellationQueued,
+      StudentAssessmentStatus.Cancelled,
+    ].includes(studentAssessment.studentAssessmentStatus);
+    const auditUser = this.systemUsersService.systemUser;
+    const now = new Date();
+    if (workflowIsCancelled) {
+      if (!options?.workflowData) {
+        // No update is executed because the data was already present or the workflow was cancelled.
+        return false;
```

Comment:
May or may not be part of the PR. But this logic should be part of the controller rather than service.

Reply:
It may, but the controller is already executing too much and I was not focusing on going further in any refactoring.
The controller logic should be segregated already in at least two other methods but I have no intention to address this as part of this ticket goal 😉